packages/serverpod_test/lib/src/serverpod_test.dart

```diff
@@ -0,0 +1,415 @@
+import 'dart:async';
+
+import 'package:serverpod/serverpod.dart';
+import 'package:test/test.dart';
+
+export 'package:meta/meta.dart' show isTestGroup;
+
+/// A test specific session that is used to call database methods or pass to endpoints.
+abstract class TestSession implements DatabaseAccessor {
+  /// AuthenticationInfo for the session.
+  AuthenticationInfo? get authenticationInfo;
+
+  /// Creates a new unique session with the provided properties.
+  /// This is useful for setting up different session states in the tests
+  /// or simulating multiple users.
+  Future<TestSession> copyWith({
+    AuthenticationInfo? Function()? getAuthenticationInfo,
+    bool? enableLogging,
+  });
+}
+
+/// Internal test session that contains implementation details that should
+/// only be used internally by the test tools.
+class InternalTestSession extends TestSession {
+  final List<InternalTestSession> _allTestSessions;
+  final TestServerpod _testServerpod;
+  AuthenticationInfo? _authenticationInfo;
+  final bool _enableLogging;
+
+  /// The underlaying Serverpod session
+  late Session serverpodSession;
+
+  @override
+  Database get db => serverpodSession.db;
+
+  @override
+  AuthenticationInfo? get authenticationInfo => _authenticationInfo;
+
+  @override
+  Transaction? transaction;
+
+  /// Creates a new internal test session.
+  InternalTestSession(
+    TestServerpod testServerpod, {
+    AuthenticationInfo? authenticationInfo,
+    InternalTestSession? sessionWithDatabaseConnection,
+    this.transaction,
+    required bool enableLogging,
+    required List<InternalTestSession> allTestSessions,
+  })  : _allTestSessions = allTestSessions,
+        _authenticationInfo = authenticationInfo,
+        _testServerpod = testServerpod,
+        _enableLogging = enableLogging {
+    _allTestSessions.add(this);
+  }
+
+  /// Sets the underlaying Serverpod session and configures it.
+  /// This is needed to enable the creation of the InternalTestSession before
+  /// the Serverpod session is created.
+  void setAndConfigureServerpodSession(Session session) {
+    serverpodSession = session;
+
+    serverpodSession.transaction = transaction;
+
+    serverpodSession.updateAuthenticated(_authenticationInfo);
+  }
+
+  @override
+  Future<TestSession> copyWith({
+    AuthenticationInfo? Function()? getAuthenticationInfo,
+    bool? enableLogging,
+  }) async {
+    var newSession = InternalTestSession(
+      _testServerpod,
+      allTestSessions: _allTestSessions,
+      authenticationInfo: getAuthenticationInfo != null
+          ? getAuthenticationInfo()
+          : _authenticationInfo,
+      enableLogging: enableLogging ?? _enableLogging,
+      transaction: transaction,
+    );
+    var newServerpodSession = await _testServerpod.createSession(
+      enableLogging: enableLogging ?? _enableLogging,
+    );
+
+    newSession.setAndConfigureServerpodSession(newServerpodSession);
```

Comment:
Question(For my understanding): The reason this can not be done in the constructor is because we preserve all test sessions between the tests, since we don't control their environment, but reset their internal session?

Reply:
That is one reason. Although the original reason was that the user-provided `testClosure` which takes the `InternalTestSession` has to be called inside our test `group`. Creating a Serverpod session is an `async` call, but test `group`s are not allowed to be async so that has to happen in the `setUpAll`. 

Maybe this could be done differently, but that seemed the most straight forward way!